### DISCUSSION TITLE
Runge-Kutta 4/5 scheme requirement for extra Variable ‘next_dt’`

### DISCUSSION THREAD

**Discussion by selipot** [Parcels-code/Parcels#1534] [Q&A]
I would like to test the Runge-Kutta 4 advection scheme against the Runge-Kutta 4/5 scheme. The docstring for the kernel AdvectionRK45 states that:

`Note that this kernel requires a Particle Class that has an extra Variable ‘next_dt’` 

Is there more information in the documentation about what that extra variable corresponds to? I am unable to find it. Any help or pointer would be greatly appreciated.

**Comment by erikvansebille** [comment]
HI @selipot, you can see an example of how that's done (using parcels v3) below:

```python
RK45Particle = JITParticle.add_variable('next_dt', dtype=np.float32, initial=dt) 
```
And then use `RK45Particle` instead of `JITParticle` in your `pset` creation

For older Parcels versions, you don't need to set this `next_dt` explicitly

Note that @daanreijnders found an issue with the RK45 implementation at https://github.com/OceanParcels/parcels/issues/1516, that's on my list to explore/fix in the coming days. So keep an eye on that Issue

**Comment by selipot** [comment]
Hi @erikvansebille thank you for your reply. I am currently using version 2.4.1.dev9 so if I understand properly what you wrote I should simply define my kernel as `kernel = pset.Kernel(AdvectionRK45)`  or is it something else? The same code works with `kernel = pset.Kernel(AdvectionRK4)` but fails with `AdvectionRK45`. I am getting a `zarr.errors.BoundsCheckError: index out of bounds for dimension with length 2` 

(I'd like to take note of the issue that @daanreijnders posted at https://github.com/OceanParcels/parcels/issues/1516 but the link to the notebook there seems outdated.)

**Reply by selipot** [reply]
Actually, stay tuned: I simplified my code to *not* use parallel computation with mpi and it seems to run. I will let you know if I get into errors like what @daanreijnders noticed.

**Reply by daanreijnders** [reply]
Hi Sean, how did your runs go?

**Reply by selipot** [reply]
My single global run actually went fine and I did not encounter the type of issues you raised in https://github.com/OceanParcels/parcels/issues/1516. I was able to compare this run to an identical run using RK4, in terms of particle velocity variance in Eulerian bins and only found noticeable, yet modest, differences in regions of high eddy kinetic energy of this global high-resolution model. Other more advanced diagnostics also revealed very little difference (manuscript in prep!).

**Reply by daanreijnders** [reply]
Hmm, that is interesting. Did you play around with the tolerance? The culprit was that Parcels didn't automatically convert the tolerance from degrees to meters, so in spherical setups, it allowed for very large errors. Erik's fixed this in #1539.

**Reply by selipot** [reply]
I did not change the tolerance at all. Maybe the fact that my velocity field was at approximately 1/25 degree resolution helped?

**Reply by daanreijnders** [reply]
I think the resolution wouldn't have influenced this. The bug was in the tolerance being read in degrees rather than meters, which was determined by whether the data was provided in spherical or flat coordinates (the latter being the case for the unit tests, which is why the issue wasn't spotted before).

I'm a bit puzzled as to why you didn't experience this issue at all. @erikvansebille you verified the results of my run too, right? 
It might be worthwhile double checking your run with RK45 in the latest version of parcels, for just 10 days or so...

**Reply by erikvansebille** [reply]
> @erikvansebille you verified the results of my run too, right?

No, I didn't rerun your simulations from #1516. I wrote some unit tests to verify if the new code worked as expected, but haven't tested it in the wild. 
The link to the zip you prepared has expired, can you try to rerun it, @daanreijnders?

**Reply by daanreijnders** [reply]
I reran my test with the new changes and it was fine, though much slower than RK4.

What I meant to ask was whether you reproduced the error I had before the fix was implemented. I just find it strange that @selipot didn't run into it, so I'm now doubting whether something else was wrong with [my setup](https://filesender.surf.nl/?s=download&token=743a49dd-e448-41cf-b99d-ff71436b245c).
